lib/semantics/check-io.cc

```diff
@@ -15,12 +15,62 @@
 #include "check-io.h"
 #include "expression.h"
 #include "tools.h"
+#include "../common/format.h"
 #include "../parser/tools.h"
 
 namespace Fortran::semantics {
 
 // TODO: C1234, C1235 -- defined I/O constraints
 
+void IoChecker::Enter(
+    const parser::Statement<common::Indirection<parser::FormatStmt>> &stmt) {
+  if (!stmt.label.has_value()) {
+    context_.Say("Format statement must be labeled"_err_en_US);  // C1301
+  }
+  parser::CharBlock stmtCharBlock{stmt.source};
+  auto parenIndex{stmtCharBlock.ToString().find('(')};
+  parser::CharBlock formatCharBlock{parser::CharBlock(
+      stmtCharBlock.begin() + parenIndex, stmtCharBlock.size() - parenIndex)};
+  static constexpr int maxFormatErrors{3};
+  common::FormatError formatErrors[maxFormatErrors];
+  int errorCount{0};
+  switch (context_.GetDefaultKind(TypeCategory::Character)) {
+  case 1: {
+    common::FormatValidator<char> validator{formatCharBlock.begin(),
+        formatCharBlock.size(), IoStmtKind::None,
+        context_.warnOnNonstandardUsage(), formatErrors, maxFormatErrors};
+    errorCount = validator.Check();
+    break;
+  }
+  case 2: {
+    // TODO: Get this to work.
+    common::FormatValidator<char16_t> validator{/*???*/ nullptr,
+        /*???*/ formatCharBlock.size(), IoStmtKind::None,
+        context_.warnOnNonstandardUsage(), formatErrors, maxFormatErrors};
+    errorCount = validator.Check();
+    break;
+  }
+  case 4: {
+    // TODO: Get this to work.
+    common::FormatValidator<char32_t> validator{/*???*/ nullptr,
+        /*???*/ formatCharBlock.size(), IoStmtKind::None,
+        context_.warnOnNonstandardUsage(), formatErrors, maxFormatErrors};
+    errorCount = validator.Check();
+    break;
+  }
+  default: CRASH_NO_CASE;
+  }
+  for (int i{0}; i < errorCount; ++i) {
+    common::FormatError *error{formatErrors + i};
+    parser::MessageFormattedText msg{
+        parser::MessageFixedText(error->text_, strlen(error->text_)),
+        error->arg_};
+    parser::CharBlock errorCharBlock{parser::CharBlock(
+        formatCharBlock.begin() + error->offset_, error->length_)};
+    context_.Say(errorCharBlock, "%s"_err_en_US, msg.string());
```

Comment:
`Say(errorCharBlock, msg)` should be good enough if `isFatal` is set correctly in msg.

Reply:
Done; thanks.